### PR TITLE
Augur Updates for RSV-A and RSV-B

### LINKED ISSUE #481: Add RSV to the Augur workflow
:cool:

### :pushpin: Explain the Request

At present, Augur only works with SARS-CoV-2, Mpox and seasonal influenza.

RSV is an important respiratory pathogen, similar to flu, whose addition to the workflow will greatly benefit public health labs understanding of its genomic epidemiology.

### PR DESCRIPTION
**Pull request by jrotieno**
This PR closes #481.

🗑️ This dev branch should <NOT> be deleted after merging to main.

## :brain: Aim, Context and Functionality

This PR enables the Augur workflow to be able to analyze RSV-A and RSV-B assemblies. Prior to this PR, the workflow would only process SARS-CoV-2, seasonal Influenza A and B, and Mpox.

The Augur Prep workflow has not been updated as there was no need for changes therein.

## :hammer_and_wrench:  Impacted Workflows/Tasks & Changes Being Made
This will affect the behavior of the workflow(s) even if users don’t change any workflow inputs relative to the last version : Yes

Previously, for seasonal Influenza and Mpox, augur reference files and default values were set in the [augur_utilities](https://github.com/theiagen/public_health_bioinformatics/blob/main/tasks/utilities/data_handling/task_augur_utilities.wdl) task. With the move to house all organism specific defaults to the [organism_parameters](https://github.com/theiagen/public_health_bioinformatics/blob/main/workflows/utilities/wf_organism_parameters.wdl) workflow, we have now moved the default augur parameters for seasonal Influenza and Mpox to the organism_parameters workflow.

SARS-CoV-2 augur defaults have not been moved yet as the current behaviour is to download the augur defaults directly from nextstrain's "https://github.com/nextstrain/ncov" page. As SARS-CoV-2 clades tsv file is updated quite frequently with new lineages, this is the best strategy. In the future, the other pathogens may also take this direction.

Running this workflow on different occasions could result in different results, e.g. due to use of a live database, "latest" docker image, or stochastic data processing : Yes, if the augur reference files are updated in the future.

#### 🔄 Data Processing 

Docker/software or software versions changed: No

Databases or database versions changed: No

Data processing/commands changed: Yes, organism default sources moved.

File processing changed: Yes, organism default sources moved.

Compute resources changed: No

#### ➡️ Inputs 

Exposed the following optional inputs, used in augur_frequencies, but not currently used in the workflow:
```
augur	min_date	Float
augur	pivot_interval	Int
augur	proportion_wide	Float
augur	narrow_bandwidth	Float
```
The following optional inputs have been removed:
```
flu_defaults	cpu	Int
flu_defaults	disk_size	Int
flu_defaults	docker	String
flu_defaults	flu_lat_longs_tsv	File
flu_defaults	memory	Int
mpxv_defaults	cpu	Int
mpxv_defaults	disk_size	Int
mpxv_defaults	docker	String
mpxv_defaults	memory	Int
mpxv_defaults	mpxv_auspice_config	File
mpxv_defaults	mpxv_clades_tsv	File
mpxv_defaults	mpxv_lat_longs_tsv	File
mpxv_defaults	mpxv_reference_fasta	File
mpxv_defaults	mpxv_reference_genbank	File
```
The following optional inputs are exposed in the workflow but not used for this workflow:
```
organism_parameters	gene_locations_bed_file	File
organism_parameters	genome_length_input	Int
organism_parameters	hiv_primer_version	String
organism_parameters	kraken_target_organism_input	String
organism_parameters	nextclade_dataset_name_input	String
organism_parameters	nextclade_dataset_tag_input	String
organism_parameters	pangolin_docker_image	String
organism_parameters	primer_bed_file	File
organism_parameters	reference_gff_file	File
organism_parameters	vadr_max_length	Int
organism_parameters	vadr_mem	Int
organism_parameters	vadr_options	String
organism_parameters	vadr_skip_length	Int
```

#### ⬅️ Outputs 

None changed

#### Commandline Testing with MiniWDL or Cromwell (optional)

Locally tested using the RSV-A dataset, see Terra testing below, and works as expected. 

#### Terra Testing

As we moved the augur pathogen specific parameters to the organism parameters workflow, we tested all previous and the new pathogens to ensure that those still work.

RSV-A: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/8e750146-a665-4aff-ac35-ca6571c51cd3
RSV-B: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/e58e51e0-267c-4e0c-a9c2-ec01154dfaef
H1N1 HA: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/47d664d9-2b13-42d8-b888-e04db1fe2b0b
H1N1 NA: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/5335d341-30cb-49a0-8373-a291151e4d2e
H3N2 HA: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/471a1ff5-51fb-46df-8656-52ac6bd9f0e3
H3N2 NA: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/90c8555e-c08e-415c-9c8f-4874baf74cdc
MPOX: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/83859995-c15a-4606-972c-7568ad9c59f8
SARS-COV-2: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/f9b4e5bc-57e4-4474-a448-31b00d3fa45c

#### Suggested Scenarios for Reviewer to Test

It was noticed that the augur_refine step fails when the date information does not align with the genetic divergence. For example, the RSV-A sequences were initially given dates in 2023 and would fail this step. However, when the dates were corrected to the appropriate collection dates for the samples in 2013-2014, the workflow was able to run to completion successfully.
successful run: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/fa391728-0fa6-448f-9c4a-64cddc60c9ae
failed run: https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Otieno_Sandbox/job_history/0b37dff6-5ee5-4bb5-ab11-285018de7ba0

The reviewer can test any additional appropriately dated sequences to see if there are any failures.

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested locally and results, including file contents, are as anticipated
- [x] The workflow/task has been tested on Terra and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (to be completed by Theiagen developer)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

## 🎯 Reviewer Checklist 

- [x] All impacted workflows/tasks have been tested on Terra with a different dataset than used for development
- [x] All reviewer-suggested scenarios have been tested and any additional
- [x] All changed results have been confirmed to be accurate
- [x] All workflows/tasks impacted by change/s have been tested using a standard validation dataset to ensure no unintended change of functionality
- [x] All code adheres to the style guide
- [x] MD5 sums have been updated
- [x] The PR author has addressed all comments

## 🗂️ Associated Documentation (to be completed by Theiagen developer)

- [ ] Relevant documentation on the Public Health Resources "PHB Main" has been updated
- [x] Workflow diagrams have been updated to reflect changes

### CODE REVIEW

File: workflows/phylogenetics/wf_augur.wdl [hunk 1]
```diff
@@ -150,8 +152,8 @@ workflow augur {
                             augur_clades.clade_assignments_json,
                             augur_traits.traits_assignments_json]),
         build_name = build_name,
-        lat_longs_tsv = select_first([sc2_defaults.lat_longs_tsv, flu_defaults.lat_longs_tsv, mpxv_defaults.lat_longs_tsv, lat_longs_tsv]),
-        auspice_config = select_first([sc2_defaults.auspice_config, flu_defaults.auspice_config, mpxv_defaults.auspice_config, auspice_config])
+        lat_longs_tsv = select_first([organism_parameters.augur_lat_longs_tsv, sc2_defaults.lat_longs_tsv, lat_longs_tsv]),
+        auspice_config = select_first([organism_parameters.augur_auspice_config, sc2_defaults.auspice_config, auspice_config])
```
Comment (kapsakcj) [inline workflows/phylogenetics/wf_augur.wdl:156]: change priority - user-provided inputs to beginning, then SC2_defaults, then organism_parameters

Comment (jrotieno) [reply workflows/phylogenetics/wf_augur.wdl]: Done

File: workflows/utilities/wf_organism_parameters.wdl [hunk 2]
```diff
@@ -49,6 +60,19 @@ workflow organism_parameters {
...
+    #File sc2_lat_longs_tsv = "gs://theiagen-public-files-rp/terra/flu-references/lat_longs.tsv"
+    #File sc2_clades_tsv = "gs://theiagen-public-files-rp/terra/sars-cov-2-files/sc2_clades.tsv"
+    #File sc2_reference_gbk = "gs://theiagen-public-files-rp/terra/sars-cov-2-files/MN908947.3.gb"
+    #File sc2_auspice_config = "gs://theiagen-public-files-rp/terra/sars-cov-2-files/sc2_auspice_config.json"
+    #Int sc2_min_num_unambig = 27000
+    #Float min_date = 2020.0
+    #Int pivot_interval = 1
+    #String pivot_interval_units = "weeks"
+    #Float narrow_bandwidth = 0.05
+    #Float proportion_wide = 0.0
```
Comment (kapsakcj) [inline workflows/utilities/wf_organism_parameters.wdl:74]: Do we intend to keep these commented out lines 64-74? Will they be useful in the future?

I saw the note about SC2 nextstrain files being frequently updated (like the clades.tsv file) so I wasn't sure if we should delete these lines or not.

If we don't intend to keep them for future use, could you please delete these lines?

Comment (jrotieno) [reply workflows/utilities/wf_organism_parameters.wdl:74]: Only if we plan to use augur_frequencies, task present but currently not used. Removing for SARS-CoV-2 till when needed for a cleaner look. These were set for the other pathogens, but also not used presently.

**Review by kapsakcj — APPROVED** [review]

**Comment by kapsakcj** [comment]
Apologies for the delay on reviewing, this took quite some time to test across the different organisms and troubleshoot when user error (👨 ➡️ 🪞) occurred. Thank you for your patience.

I have no further code changes to suggest or request ✅ 

Thank you for updating the documentation in Notion 👍 

Here are my test workflows and a short description of observed behavior:

- sars-cov-2: 
  - augur_prep ✅ : https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/c7aded86-d835-49bc-88c0-597373648d7d
  - Augur ✅ : https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/c7aded86-d835-49bc-88c0-597373648d7d
  - auspice JSON viewable without issues ✅ Can color tree nodes by both nextclade clade and pango lineage
- Mpox:
  - augur_prep ✅ https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/fde787cb-eda8-4b8b-9300-536396a7ff68
  - augur ✅ https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/242fc2c6-dfad-4002-b565-62332bdb6e7e
  - notes: auspice_input_json shows clade_membership is unassigned for all samples. But we can overlay the merged-metadata.tsv file to color by clades in auspice
- Flu:
  - augur_prep ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/49fa61f1-c4e3-43b0-a488-2ad7d19187ab
  - augur ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/1950f4c9-1745-4c79-b8b5-693471dec9b0
  - notes: H3N2 samples. Was able to visualize the auspice_input_json and color clades after overlaying merged-metadata.tsv file. 💡 Reminder to users: The user should set `organism` = `flu` and set the `flu_segment` to `HA` or `NA` and additionally set `flu_subtype` to `H1N1` or `H3N2`. This was not immediately clear to me and I spent time troubleshooting.
- rsv-a:
  - augur_prep ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/61b94c5b-76dc-4af5-81de-e7ebde9d0288
  - augur ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/ab12c336-1e93-4fb0-800c-eb18cb42a6a5
  - notes: auspice_input_json shows clade_membership for 1 sample included. 2 samples pruned during wf (I'm not terribly surprised based on my poor choice of test samples). But can always overlay clade membership via meatdata TSV
- rsv-b:
  - augur_prep ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/d943bb81-3165-4b24-8f2e-5f58da352581
  - augur ✅: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/0df3e391-4530-47b0-a658-6ce7907781fb
  - notes: Ran successfully, but only after setting `distance_tree_only` to true. I think it failed previously on augur_refine step because of the very odd makeup of the samples (and their respective qualities) I included in my set. What matters is that the workflow ran, and used the reference materials that are now set by default in the organism parameters subworkflow.

**Pull request merged**